File: app/src/main/java/com/petnbu/petnbu/api/StorageApi.java
package com.petnbu.petnbu.api;


import android.graphics.Bitmap;
import android.net.Uri;
import android.util.Log;

import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;
import com.petnbu.petnbu.BuildConfig;

import java.io.ByteArrayOutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import timber.log.Timber;

/**
 * Created by Quang Quang on 11/23/2016.
 */

public class StorageApi {

    public static final String TAG = StorageApi.class.getSimpleName();

    // map files to file

    public static void updateImage(Uri uri, String fileName, OnResultListener onResultListener) {
        Log.i(TAG, "updateImage: update file " + uri.toString());
        StorageReference ref = getStorageRef().child("image_" + System.currentTimeMillis() + fileName);
        ref.putFile(uri)
                .addOnSuccessListener(taskSnapshot -> onResultListener.onSuccess(fileName, taskSnapshot))
                .addOnFailureListener(e -> onResultListener.onFailure(fileName, e));
    }

    public static void updateBitmap(Bitmap bitmap, String fileName, OnResultListener onResultListener) {
        Timber.i("updateBitmap " + fileName);
        StorageReference ref = getStorageRef().child("image_" + System.currentTimeMillis() + fileName);
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        bitmap.compress(Bitmap.CompressFormat.JPEG, 75, outputStream);
        byte[] data = outputStream.toByteArray();
        ref.putBytes(data)
                .addOnSuccessListener(taskSnapshot -> onResultListener.onSuccess(fileName, taskSnapshot))
                .addOnFailureListener(e -> onResultListener.onFailure(fileName, e));
    }

    public static StorageReference getStorageRef() {
        StorageReference storage = FirebaseStorage.getInstance().getReferenceFromUrl(BuildConfig.FIREBASE_STORAGE_URL);
        return storage.child("pets").child("photos");
    }

    public static abstract class OnUploadingImage {

        private AtomicInteger mTotalImage;
        private List<String> mList;
        private List<String> mResult;
        private OnResultListener mOnResultListener = new OnResultListener() {
            @Override
            public void onSuccess(String fileName, UploadTask.TaskSnapshot taskSnapshot) {
                mTotalImage.decrementAndGet();
                mResult.add(taskSnapshot.getDownloadUrl().toString());
                if (mTotalImage.get() == 0) {
                    onCompleted(mResult);
                }
            }

            @Override
            public void onFailure(String fileName, Exception e) {
                mTotalImage.decrementAndGet();
                onFailed(e);
            }
        };

        public OnUploadingImage(List<String> imageList) {
            mTotalImage = new AtomicInteger(imageList.size());
            mList = imageList;
            mResult = new ArrayList<>();
        }

        public void start() {
            for (String file : mList) {
                Uri uri = Uri.parse(file);
                updateImage(uri, uri.getLastPathSegment(), mOnResultListener);
            }
        }

        public abstract void onCompleted(List<String> result);

        public abstract void onFailed(Exception e);
    }

    public interface OnResultListener {

        void onSuccess(String fileName, UploadTask.TaskSnapshot taskSnapshot);

        void onFailure(String fileName, Exception e);
    }
}

File: app/src/main/java/com/petnbu/petnbu/api/StorageApi.kt
package com.petnbu.petnbu.api


import android.graphics.Bitmap
import android.net.Uri
import androidx.core.net.toUri
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.google.firebase.storage.UploadTask
import com.petnbu.petnbu.BuildConfig
import timber.log.Timber
import java.io.ByteArrayOutputStream
import java.util.concurrent.atomic.AtomicInteger

/**
 * Created by Quang Quang on 11/23/2016.
 */

object StorageApi {

    val storageRef: StorageReference
        get() {
            val storage = FirebaseStorage.getInstance().getReferenceFromUrl(BuildConfig.FIREBASE_STORAGE_URL)
            return storage.child("pets").child("photos")
        }

    // map files to file

    fun updateImage(uri: Uri, fileName: String?, onResultListener: OnResultListener) {
        Timber.i("updateImage: update file %s", uri.toString())
        val ref = storageRef.child("image_" + System.currentTimeMillis() + fileName)
        ref.putFile(uri)
                .addOnSuccessListener { taskSnapshot -> onResultListener.onSuccess(fileName, taskSnapshot) }
                .addOnFailureListener { e -> onResultListener.onFailure(fileName, e) }
    }

    fun updateBitmap(bitmap: Bitmap, fileName: String, onResultListener: OnResultListener) {
        Timber.i("updateBitmap $fileName")
        val ref = storageRef.child("image_" + System.currentTimeMillis() + fileName)
        val outputStream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, 75, outputStream)
        val data = outputStream.toByteArray()
        ref.putBytes(data)
                .addOnSuccessListener { taskSnapshot -> onResultListener.onSuccess(fileName, taskSnapshot) }
                .addOnFailureListener { e -> onResultListener.onFailure(fileName, e) }
    }

    abstract class OnUploadingImage(private val mList: List<String>,
                                    private val mTotalImage: AtomicInteger = AtomicInteger(mList.size),
                                    private val mResult: MutableList<String> = ArrayList(mList.size)) {

        private val mOnResultListener = object : OnResultListener {
            override fun onSuccess(fileName: String?, taskSnapshot: UploadTask.TaskSnapshot) {
                mTotalImage.decrementAndGet()
                mResult.add(taskSnapshot.downloadUrl!!.toString())
                taskSnapshot.downloadUrl?.let { mResult.add(it.toString()) }
                if (mTotalImage.get() == 0) {
                    onCompleted(mResult)
                }
            }

            override fun onFailure(fileName: String?, e: Exception) {
                mTotalImage.decrementAndGet()
                onFailed(e)
            }
        }


        fun start() {
            for (filePath in mList) {
                val fileUri = filePath.toUri()
                updateImage(fileUri, fileUri.lastPathSegment, mOnResultListener)
            }
        }

        abstract fun onCompleted(result: List<String>)

        abstract fun onFailed(e: Exception)
    }

    interface OnResultListener {

        fun onSuccess(fileName: String?, taskSnapshot: UploadTask.TaskSnapshot)

        fun onFailure(fileName: String?, e: Exception)
    }
}
